### PR TITLE
[Security] Add Callable voter

### PR DESCRIPTION
**Pull request by instabledesign**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

Simple implementation of a CallableVoter.

The `callable` must return true, false, mixed to respectively grant, denied, abstain.

It allow this multiple syntax use

```php
use Symfony\Component\Security\Core\Authorization\Voter\CallableVoter;

new CallableVoter(
    function () {
        return true;
    }
);

new CallableVoter(
    [
        new class()
        {
            public function get()
            {
                return true;
            }
        },
        'get',
    ]
);

new CallableVoter(
    new class()
    {
        public function __invoke()
        {
            return true;
        }
    }
);
```

What did you think about such addition?
Feel free to comment or close it.

### CODE REVIEW

**Comment by iltar** [comment]
Do you have any scenarios where you'd like to use this over the traditional style? It adds a layer of complexity in my eyes, that I'm not yet sure of that we need.

**Comment by instabledesign** [comment]
You're right. I'll create `CustomVoter` when i need it to.
To be honest my current case was a demo code to understand quickly what is going on during the decision process

```php
$accessDecisionManager = new AccessDecisionManager([
    new CallableVoter(function (TokenInterface $token, $subject, $attributes) {
        if ($token->getUsername() === 'admin') {
            return true;
        }

        return null;
    }),
    new CallableVoter(function (TokenInterface $token, $subject, $attributes) {
        return //other test cases;
    }),
]);
```

**Comment by instabledesign** [comment]
rerun tests

File: src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authorization\Voter;
+
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+
+/**
+ * CallableVoter delegate vote to the callable.
+ * Callable received the token, the subject and the attributes
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php]: missing blank lie to separate title from description

File: src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        $this->callable = $callable;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function vote(TokenInterface $token, $subject, array $attributes)
+    {
+        $result = call_user_func_array($this->callable, array($token, $subject, $attributes));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php]: no need for cal_user_func_array:
`$result = ($this->callable)($token, $subject, $attributes);`

File: src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+class CallableVoterTest extends TestCase
+{
+    /**
+     * @dataProvider getVoteTests
+     */
+    public function testVote($callable, $expected)
+    {
+        $voter = new CallableVoter($callable);
+        $tokenMock = $this->getMockBuilder('Symfony\Component\Security\Core\Authentication\Token\TokenInterface')->getMock();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php]: should be `TokenInterface::class`

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    {
+        return array(
+            array(
+                function () {
+                    return true;
+                },
+                Voter::ACCESS_GRANTED,
+            ),
+            array(
+                \Closure::fromCallable(function () {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php]: why wrapping in `Closure::fromCallable`?

Comment (instabledesign) [reply src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php]: to test with `\Closure` object.
You think it's useless?

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php]: It's useless then, yes: a closure is already a `Closure`

Comment (instabledesign) [reply src/Symfony/Component/Security/Core/Tests/Authorization/Voter/CallableVoterTest.php]: You got me 😆...

File: src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function vote(TokenInterface $token, $subject, array $attributes)
+    {
+        $result = ($this->callable)($token, $subject, $attributes);
+
+        if (false === $result) {
+            return Voter::ACCESS_DENIED;
```
Comment (derrabus) [inline src/Symfony/Component/Security/Core/Authorization/Voter/CallableVoter.php]: Those constants are defined on the `VoterInterface`, `Voter` just inherits them from there. I'd prefer to use `VoterInterface::ACCESS_DENIED` here.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by instabledesign** [comment]
it's ok for me and test green
@derrabus @nicolas-grekas

**Comment by lyrixx** [comment]
I'm not sure this class is really useful. More over, you said you used it only for test purpose.
Instead you could you anonymous classes:

```
$voter = new class() implements VoterInterface{
    public function vote(...);
};
```

Is almost the same, isn't?

**Comment by instabledesign** [comment]
Yes you're right
> my case was php < 7
did you think this addition has his place in older version?

**Comment by stof** [comment]
I would vote -1 on this PR.

The only difference between the CallableVoter using an inline callback and the anonymous class implementing VoterInterface is that CallableVoter expects to return `boolean|mixed` instead of using the 3 states of the vote, and performs the conversion on its own. This seems like a drawback to me, as it creates 2 different signatures for the voting.

Thus, such voter would be almost useless in a fullstack project (you cannot use a closure there to configure the voter, so you would still have to build the callable in a class, and then implementing VoterInterface directly gives you the benefits of autoconfiguration).

**Comment by lyrixx** [comment]
I agree with @stof

**Comment by chalasr** [comment]
Same for me. 
Closing as 3 core members voted -1, thank you for proposing.

**Pull request closed without merging**